File: unity-package/Editor/Tileable/OffsetWrap.cs
using System;
using UnityEngine;

namespace UnityAiAssets.Editor.Tileable
{
    /// <summary>
    /// Circular offset / wrap helpers shared by previews and correction.
    /// </summary>
    public static class OffsetWrap
    {
        public static int WrapCoordinate(int value, int size)
        {
            if (size <= 0) throw new ArgumentOutOfRangeException(nameof(size));
            var wrapped = value % size;
            return wrapped < 0 ? wrapped + size : wrapped;
        }

        public static void OffsetShiftAmounts(int width, int height, out int dx, out int dy,
            float fraction = SeamThresholds.OffsetPreviewFraction)
        {
            if (width <= 0 || height <= 0) throw new ArgumentOutOfRangeException();
            dx = Mathf.RoundToInt(width * fraction) % width;
            dy = Mathf.RoundToInt(height * fraction) % height;
        }

        public static Color32[] CircularShift(Color32[] source, int width, int height, int dx, int dy)
        {
            if (source == null) throw new ArgumentNullException(nameof(source));
            if (source.Length != width * height) throw new ArgumentException("Pixel buffer size mismatch.");
            dx = WrapCoordinate(dx, width);
            dy = WrapCoordinate(dy, height);
            var result = new Color32[source.Length];
            for (var y = 0; y < height; y++)
            {
                for (var x = 0; x < width; x++)
                {
                    var sx = WrapCoordinate(x - dx, width);
                    var sy = WrapCoordinate(y - dy, height);
                    result[y * width + x] = source[sy * width + sx];
                }
            }

            return result;
        }

        public static Color32[] OffsetPreview(Color32[] source, int width, int height)
        {
            OffsetShiftAmounts(width, height, out var dx, out var dy);
            return CircularShift(source, width, height, dx, dy);
        }

        public static Color32[] TiledPreview(Color32[] source, int width, int height,
            int repeats = SeamThresholds.DefaultTilePreviewRepeat)
        {
            if (repeats < 1) throw new ArgumentOutOfRangeException(nameof(repeats));
            var outW = width * repeats;
            var outH = height * repeats;
            var result = new Color32[outW * outH];
            for (var ty = 0; ty < repeats; ty++)
            {
                for (var tx = 0; tx < repeats; tx++)
                {
                    for (var y = 0; y < height; y++)
                    {
                        for (var x = 0; x < width; x++)
                        {
                            result[(ty * height + y) * outW + (tx * width + x)] = source[y * width + x];
                        }
                    }
                }
            }

            return result;
        }
    }
}

File: src/unity_ai_assets/processing/offset_wrap.py
"""Circular offset / wrap helpers shared by previews, analysis, and correction."""

from __future__ import annotations

from PIL import Image

from unity_ai_assets.processing.seam_thresholds import (
    DEFAULT_TILE_PREVIEW_REPEAT,
    OFFSET_PREVIEW_FRACTION,
)


def wrap_coordinate(value: int, size: int) -> int:
    """Wrap an integer coordinate into ``[0, size)``."""
    if size <= 0:
        raise ValueError("size must be positive")
    return value % size


def offset_shift_amounts(
    width: int,
    height: int,
    *,
    fraction: float = OFFSET_PREVIEW_FRACTION,
) -> tuple[int, int]:
    """Return (dx, dy) pixel shifts for an offset/seam-centering preview."""
    if width <= 0 or height <= 0:
        raise ValueError("width and height must be positive")
    dx = int(round(width * fraction)) % width
    dy = int(round(height * fraction)) % height
    return dx, dy


def circular_shift(image: Image.Image, dx: int, dy: int) -> Image.Image:
    """Shift pixels by ``(dx, dy)`` with toroidal wrapping (no empty borders).

    Positive ``dx`` moves content right; positive ``dy`` moves content down.
    The returned image is always a new copy. Source pixel ``(x, y)`` lands at
    ``((x + dx) % width, (y + dy) % height)``.
    """
    width, height = image.size
    if width == 0 or height == 0:
        raise ValueError("image must have positive dimensions")

    dx = wrap_coordinate(dx, width)
    dy = wrap_coordinate(dy, height)
    if dx == 0 and dy == 0:
        return image.copy()

    # Horizontal wrap: dest[x] comes from source[(x - dx) % w].
    if dx == 0:
        horizontal = image.copy()
    else:
        horizontal = Image.new(image.mode, (width, height))
        left_src = image.crop((width - dx, 0, width, height))
        right_src = image.crop((0, 0, width - dx, height))
        horizontal.paste(left_src, (0, 0))
        horizontal.paste(right_src, (dx, 0))

    if dy == 0:
        return horizontal

    # Vertical wrap: dest[y] comes from source[(y - dy) % h].
    result = Image.new(image.mode, (width, height))
    top_src = horizontal.crop((0, height - dy, width, height))
    bottom_src = horizontal.crop((0, 0, width, height - dy))
    result.paste(top_src, (0, 0))
    result.paste(bottom_src, (0, dy))
    return result


def offset_preview(image: Image.Image, *, fraction: float = OFFSET_PREVIEW_FRACTION) -> Image.Image:
    """Shift by ~50% on both axes so seams move toward the center."""
    dx, dy = offset_shift_amounts(image.width, image.height, fraction=fraction)
    return circular_shift(image, dx, dy)


def tiled_preview(
    image: Image.Image,
    *,
    repeats: int = DEFAULT_TILE_PREVIEW_REPEAT,
) -> Image.Image:
    """Compose an ``repeats x repeats`` tiled view (wrap-mode preview)."""
    if repeats < 1:
        raise ValueError("repeats must be >= 1")
    width, height = image.size
    canvas = Image.new(image.mode, (width * repeats, height * repeats))
    for ty in range(repeats):
        for tx in range(repeats):
            canvas.paste(image, (tx * width, ty * height))
    return canvas
